File: 198-house-robber/house-robber.java
class Solution {
    public int rob(int[] nums) {
        if(nums.length == 1) return nums[0];
        // if(nums.length == 2) return Math.max(nums[0], nums[1]);// ? nums[0] : nums[1];

        int one = 0;
        int two = 0;

        for(int i=0;i<nums.length;i++) {
            int temp = one;
            one = Math.max(nums[i]+two, one);
            two = temp;
        }
        return Math.max(one, two);



    }
}

File: 198-house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        #--------------------------Bottom Up ------------------------------------#
        if not nums: return 0

        if len(nums) == 1: return nums[0]

        dp = [0] * len(nums)
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for i in range(2, len(nums)):
            dp[i] = max(nums[i] + dp[i-2], dp[i-1])

        return dp[-1]

        #--------------------------Top down ------------------------------------#
        # dp = [-1] * len(nums)
        # def rec(i):
        #     if i >= len(nums):
        #         return 0
            
        #     if i == len(nums) - 1: return nums[i]

        #     if dp[i] != -1: return dp[i]
            
        #     dp[i] = max(nums[i] + rec(i+2), rec(i+1))
        #     return dp[i]

        # return max(rec(0), rec(1))